File: day5/solution.py
from collections import defaultdict


def parse_stacks(text: str) -> dict[int, list[str]]:
    stacks: dict[int, list[str]] = defaultdict(list)
    for line in text.splitlines():
        if not line:
            break
        for count, letter in enumerate(line):
            if not letter.isalpha():
                continue
            col = count // 4 + 1
            stacks[col] += [letter]
    return stacks


def parse_line(line: str) -> tuple[int, int, int]:
    line = line.replace("move ", "").replace("from ", "").replace("to ", "")
    return tuple(map(int, line.split(" ")))  # type: ignore


def crate_mover_9000(text: str, stacks: dict[int, list[str]]):
    lines = text.splitlines()
    at_breakpoint = False
    for line in lines:
        if not line:
            at_breakpoint = True
            continue
        if not at_breakpoint:
            continue

        amount, f, to = parse_line(line)

        popping = stacks[f][:amount]
        del stacks[f][:amount]

        stacks[to] = list(reversed(popping)) + stacks[to]


def crate_mover_9001(text: str, stacks: dict[int, list[str]]):
    lines = text.splitlines()
    at_breakpoint = False
    for line in lines:
        if not line:
            at_breakpoint = True
            continue
        if not at_breakpoint:
            continue

        amount, f, to = parse_line(line)

        popping = stacks[f][:amount]
        del stacks[f][:amount]

        stacks[to] = popping + stacks[to]


def print_dict_list(stacks: dict[int, list[str]]):
    out = list()
    for _ in stacks.items():
        out.append("")

    for index, stack in stacks.items():
        out[index - 1] = stack[0]

    print("".join(out))


with open("input.txt", "r") as f:
    text = f.read()

stacks = parse_stacks(text)
crate_mover_9000(text, stacks)
print_dict_list(stacks)

stacks = parse_stacks(text)
crate_mover_9001(text, stacks)
print_dict_list(stacks)

File: day5/solution.hs
import Data.List (isPrefixOf, transpose)
import Data.Maybe (catMaybes, fromJust, isJust, mapMaybe)
import Text.Read (readMaybe)
import Debug.Trace (trace)

dropEnd n = reverse . drop n . reverse

takeEnd n = reverse . take n . reverse

replace list n new = left ++ [new] ++ drop 1 right
  where
    (left, right) = splitAt n list

chunksOf :: Int -> [a] -> [[a]]
chunksOf _ [] = []
chunksOf n xs = take n xs : chunksOf n (drop n xs)

parseStacks :: [String] -> [String]
parseStacks = map trim . filter isNumLine . transpose . tail . reverse
  where
    isNumLine a = not ("[" `isPrefixOf` a || " " `isPrefixOf` a || "]" `isPrefixOf` a)
    trim = takeWhile (/= ' ')

-- Count, From, To
parseInput :: String -> (Int, Int, Int)
parseInput line = (head words_, words_ !! 1 - 1, words_ !! 2 - 1)
  where
    words_ = mapMaybe readMaybe $ words line

applyMove :: ([Char] -> [Char])  -> [[Char]] -> (Int, Int, Int) -> [[Char]]
applyMove app chars (count, from, to) = new2
  where
    dropped = app $ takeEnd count (chars !! from)
    new = replace chars from (dropEnd count (chars !! from))
    new2 = replace new to ((new !! to) ++ dropped)

crateMover9000 :: [[Char]] -> [(Int, Int, Int)] -> [[Char]]
crateMover9000 = foldl (applyMove reverse)

crateMover9001 :: [[Char]] -> [(Int, Int, Int)] -> [[Char]]
crateMover9001 = foldl (applyMove id)

formatOutput :: [String] -> String
formatOutput = map last

main :: IO ()
main = do
  content <- readFile "input.txt"

  let stacks = (parseStacks . takeWhile (not . null) . lines) content
  let moves = (map parseInput . tail . dropWhile (not . null) . lines) content

  print $ formatOutput $ crateMover9000 stacks moves
  print $ formatOutput $ crateMover9001 stacks moves
